File: server/src/model_catalog.py
"""Load and validate the repository-owned model catalog."""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Any


_REPOSITORY_CATALOG = Path(__file__).resolve().parents[2] / "models" / "model-catalog.v1.json"
_PACKAGED_CATALOG = Path(__file__).resolve().parents[1] / "models" / "model-catalog.v1.json"
CATALOG_PATH = Path(
    os.getenv(
        "MODEL_CATALOG_PATH",
        str(_REPOSITORY_CATALOG if _REPOSITORY_CATALOG.is_file() else _PACKAGED_CATALOG),
    )
)


@dataclass(frozen=True)
class ModelSpec:
    """Immutable metadata used for managed model status and installation."""

    model_id: str
    label: str
    family: str
    version: str
    precision: str
    languages: tuple[str, ...]
    archive_url: str
    archive_bytes: int
    installed_bytes: int
    sha256: str
    required_files: tuple[str, ...]
    required_directories: tuple[str, ...] = ()
    role: str = ""
    runtime: str = ""
    minimum_runtime_version: str = ""
    source_url: str = ""
    weights_license: str = ""
    archive_license: str = ""
    voice_license: str = ""
    platforms: tuple[str, ...] = ()
    minimum_app_version: str = ""
    smoke_test: str = ""
    migration: str = ""
    rollback: str = ""

    @property
    def archive_size_mb(self) -> float:
        return round(self.archive_bytes / (1024 * 1024), 1)

    @property
    def installed_size_mb(self) -> float:
        return round(self.installed_bytes / (1024 * 1024), 1)

    def install_dir(self, cache_root: Path) -> Path:
        return cache_root / self.model_id


def _load_catalog() -> dict[str, Any]:
    try:
        catalog = json.loads(CATALOG_PATH.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        raise RuntimeError(f"Unable to load model catalog at {CATALOG_PATH}: {exc}") from exc

    if catalog.get("schema_version") != 1:
        raise RuntimeError("Unsupported model catalog schema")
    entries = catalog.get("models")
    if not isinstance(entries, list) or not entries:
        raise RuntimeError("Model catalog must contain a non-empty models list")

    ids = [entry.get("id") for entry in entries if isinstance(entry, dict)]
    if len(ids) != len(entries) or any(not value for value in ids) or len(set(ids)) != len(ids):
        raise RuntimeError("Model catalog IDs must be present and unique")
    return catalog


MODEL_CATALOG = _load_catalog()
MODEL_ENTRIES = {entry["id"]: entry for entry in MODEL_CATALOG["models"]}


def model_entry(model_id: str) -> dict[str, Any]:
    """Return the checked catalog entry for ``model_id``."""
    try:
        return MODEL_ENTRIES[model_id]
    except KeyError as exc:
        raise RuntimeError(f"Model is missing from the shared catalog: {model_id}") from exc


def managed_model_spec(model_id: str) -> ModelSpec:
    """Build an installer specification from one downloadable catalog entry."""
    entry = model_entry(model_id)
    archive_url = entry.get("archive_url")
    archive_bytes = entry.get("archive_bytes")
    installed_bytes = entry.get("installed_bytes")
    sha256 = entry.get("sha256")
    if not archive_url or not isinstance(archive_bytes, int) or not isinstance(installed_bytes, int) or not sha256:
        raise RuntimeError(f"Managed model metadata is incomplete: {model_id}")

    runtime = entry.get("runtime") or {}
    license_info = entry.get("license") or {}
    source = entry.get("source") or {}
    return ModelSpec(
        model_id=entry["id"],
        label=entry["label"],
        family=entry["family"],
        version=entry["version"],
        precision=entry["precision"],
        languages=tuple(entry.get("exposed_languages", [])),
        archive_url=archive_url,
        archive_bytes=archive_bytes,
        installed_bytes=installed_bytes,
        sha256=sha256,
        required_files=tuple(entry.get("required_files", [])),
        required_directories=tuple(entry.get("required_directories", [])),
        role=entry.get("role", ""),
        runtime=runtime.get("name", ""),
        minimum_runtime_version=runtime.get("minimum_version", ""),
        source_url=source.get("url", ""),
        weights_license=license_info.get("weights", ""),
        archive_license=license_info.get("archive", ""),
        voice_license=license_info.get("voices", ""),
        platforms=tuple(entry.get("platforms", [])),
        minimum_app_version=entry.get("minimum_app_version", ""),
        smoke_test=entry.get("smoke_test", ""),
        migration=entry.get("migration", ""),
        rollback=entry.get("rollback", ""),
    )


NATIVE_TTS_MODEL = managed_model_spec("kokoro-multi-lang-v1_0")
STT_MODEL = managed_model_spec("sherpa-onnx-moonshine-base-en-int8")


def is_model_complete(model_dir: Path, spec: ModelSpec = STT_MODEL) -> bool:
    """Return true only when every required model path exists and is non-empty."""
    files_complete = all(
        (model_dir / filename).is_file() and (model_dir / filename).stat().st_size > 0
        for filename in spec.required_files
    )
    directories_complete = all(
        (model_dir / dirname).is_dir() and any((model_dir / dirname).iterdir())
        for dirname in spec.required_directories
    )
    return model_dir.is_dir() and files_complete and directories_complete

File: android/app/src/main/java/com/openmobiletts/app/ModelCatalog.kt
package com.openmobiletts.app

import android.content.Context
import kotlin.math.round
import org.json.JSONObject

/** Repository-owned model metadata shared with the desktop server. */
class ModelCatalog private constructor(
    val schemaVersion: Int,
    val reviewedAt: String,
    val rawJson: String,
    private val models: Map<String, ModelSpec>,
) {
    data class ModelSpec(
        val modelId: String,
        val label: String,
        val family: String,
        val version: String,
        val precision: String,
        val role: String,
        val status: String,
        val runtime: String,
        val minimumRuntimeVersion: String,
        val sourceUrl: String,
        val weightsLicense: String,
        val archiveLicense: String,
        val voiceLicense: String,
        val supportedLanguages: List<String>,
        val verifiedLanguages: List<String>,
        val exposedLanguages: List<String>,
        val url: String,
        val archiveBytes: Long,
        val installedBytes: Long,
        val sha256: String,
        val requiredFiles: List<String>,
        val requiredDirectories: List<String>,
        val minimumAppVersion: String,
        val smokeTest: String,
        val migration: String,
        val rollback: String,
    ) {
        val archiveSizeMb: Double
            get() = round(archiveBytes.toDouble() / 1024.0 / 1024.0 * 10.0) / 10.0

        val installedSizeMb: Double
            get() = round(installedBytes.toDouble() / 1024.0 / 1024.0 * 10.0) / 10.0
    }

    fun require(modelId: String): ModelSpec = models[modelId]
        ?: error("Model is missing from the shared catalog: $modelId")

    fun managedForRole(role: String, runtime: String? = null): List<ModelSpec> =
        models.values.filter {
            it.role == role &&
                (runtime == null || it.runtime == runtime) &&
                it.url.isNotBlank() &&
                it.archiveBytes > 0L &&
                it.sha256.isNotBlank()
        }

    fun requireRole(role: String): ModelSpec {
        val matches = models.values.filter {
            it.role == role && it.url.isNotBlank() && it.archiveBytes > 0L && it.sha256.isNotBlank()
        }
        check(matches.size == 1) { "Expected one managed $role model, found ${matches.size}" }
        return matches.single()
    }

    companion object {
        private const val ASSET_NAME = "model-catalog.v1.json"

        @Volatile
        private var cached: ModelCatalog? = null

        fun load(context: Context): ModelCatalog = cached ?: synchronized(this) {
            cached ?: parse(
                context.applicationContext.assets.open(ASSET_NAME)
                    .bufferedReader()
                    .use { it.readText() },
            ).also { cached = it }
        }

        internal fun parse(rawJson: String): ModelCatalog {
            val root = JSONObject(rawJson)
            val schemaVersion = root.getInt("schema_version")
            check(schemaVersion == 1) { "Unsupported model catalog schema: $schemaVersion" }
            val array = root.getJSONArray("models")
            check(array.length() > 0) { "Model catalog is empty" }

            val parsed = LinkedHashMap<String, ModelSpec>()
            for (index in 0 until array.length()) {
                val entry = array.getJSONObject(index)
                val id = entry.getString("id")
                check(id !in parsed) { "Duplicate model catalog ID: $id" }
                val runtime = entry.getJSONObject("runtime")
                val source = entry.getJSONObject("source")
                val license = entry.getJSONObject("license")
                parsed[id] = ModelSpec(
                    modelId = id,
                    label = entry.getString("label"),
                    family = entry.getString("family"),
                    version = entry.getString("version"),
                    precision = entry.getString("precision"),
                    role = entry.getString("role"),
                    status = entry.optString("status", "stable"),
                    runtime = runtime.getString("name"),
                    minimumRuntimeVersion = runtime.getString("minimum_version"),
                    sourceUrl = source.getString("url"),
                    weightsLicense = license.getString("weights"),
                    archiveLicense = license.getString("archive"),
                    voiceLicense = license.getString("voices"),
                    supportedLanguages = entry.getJSONArray("supported_languages").toStringList(),
                    verifiedLanguages = entry.getJSONArray("verified_languages").toStringList(),
                    exposedLanguages = entry.getJSONArray("exposed_languages").toStringList(),
                    url = entry.optionalString("archive_url"),
                    archiveBytes = entry.optLong("archive_bytes"),
                    installedBytes = entry.optLong("installed_bytes"),
                    sha256 = entry.optionalString("sha256"),
                    requiredFiles = entry.getJSONArray("required_files").toStringList(),
                    requiredDirectories = entry.getJSONArray("required_directories").toStringList(),
                    minimumAppVersion = entry.getString("minimum_app_version"),
                    smokeTest = entry.getString("smoke_test"),
                    migration = entry.getString("migration"),
                    rollback = entry.getString("rollback"),
                )
            }
            return ModelCatalog(
                schemaVersion = schemaVersion,
                reviewedAt = root.getString("reviewed_at"),
                rawJson = root.toString(),
                models = parsed,
            )
        }

        private fun org.json.JSONArray.toStringList(): List<String> =
            (0 until length()).map { getString(it) }

        private fun JSONObject.optionalString(key: String): String =
            if (isNull(key)) "" else optString(key, "")
    }
}
